File: C++ Solution/MergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

class MergeSort {
public:
    vector<int> numbers;

    MergeSort(vector<int> arr) {
        numbers = arr;
    }

    void sort() {
        numbers = mergeSort(numbers);
        printArray();
    }

    vector<int> mergeSort(vector<int> arr) {
        if (arr.size() <= 1) return arr;
        int mid = arr.size() / 2;
        vector<int> left(arr.begin(), arr.begin() + mid);
        vector<int> right(arr.begin() + mid, arr.end());
        left = mergeSort(left);
        right = mergeSort(right);
        return merge(left, right);
    }

    vector<int> merge(vector<int> left, vector<int> right) {
        vector<int> result;
        int i = 0, j = 0;
        while (i < left.size() && j < right.size()) {
            if (left[i] < right[j]) result.push_back(left[i++]);
            else result.push_back(right[j++]);
        }
        while (i < left.size()) result.push_back(left[i++]);
        while (j < right.size()) result.push_back(right[j++]);
        printArray(result);
        return result;
    }

    void printArray(vector<int> arr) {
        for (int num : arr) cout << num << " ";
        cout << endl;
    }

    void printArray() {
        for (int num : numbers) cout << num << " ";
        cout << endl;
    }
};

int main() {
    vector<int> data = {38, 27, 43, 3, 9, 82, 10};
    MergeSort sorter(data);
    sorter.sort();
    return 0;
}

File: Java Solution/MergeSort.java
public class MergeSort {
    private int[] numbers;

    public MergeSort(int[] numbers) {
        this.numbers = numbers;
    }

    public void sort() {
        numbers = mergeSort(numbers);
        printArray();
    }

    private int[] mergeSort(int[] arr) {
        if (arr.length <= 1) return arr;
        int mid = arr.length / 2;
        int[] left = new int[mid];
        int[] right = new int[arr.length - mid];
        System.arraycopy(arr, 0, left, 0, mid);
        System.arraycopy(arr, mid, right, 0, arr.length - mid);
        left = mergeSort(left);
        right = mergeSort(right);
        return merge(left, right);
    }

    private int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0, k = 0;
        while (i < left.length && j < right.length) {
            if (left[i] < right[j]) result[k++] = left[i++];
            else result[k++] = right[j++];
        }
        while (i < left.length) result[k++] = left[i++];
        while (j < right.length) result[k++] = right[j++];
        printArray(result);
        return result;
    }

    public void printArray() {
        for (int num : numbers) System.out.print(num + " ");
        System.out.println();
    }

    public void printArray(int[] arr) {
        for (int num : arr) System.out.print(num + " ");
        System.out.println();
    }

    public static void main(String[] args) {
        int[] data = {38, 27, 43, 3, 9, 82, 10};
        MergeSort sorter = new MergeSort(data);
        sorter.sort();
    }
}
